File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/9f63bc7b2c354e33fad1c0c74fae046e.go
package main

import (
	"fmt"
)

type mint struct {
	v int
}

func (m mint) val() int {
	return m.v
}

func (m *mint) increment() {
	m.v++
}

func (m *mint) decrement() {
	m.v--
}

func (m *mint) add(rhs mint) {
	m.v += rhs.v
}

func (m *mint) subtract(rhs mint) {
	m.v -= rhs.v
}

func (m *mint) multiply(rhs mint) {
	m.v *= rhs.v
}

func (m *mint) divide(rhs mint) {
	m.v /= rhs.v
}

func (m mint) pow(n int) mint {
	x := m
	r := mint{v: 1}
	for n > 0 {
		if n&1 == 1 {
			r.multiply(x)
		}
		x.multiply(x)
		n >>= 1
	}
	return r
}

func (m mint) inverse() mint {
	return m.pow(m.mod() - 2)
}

func (m mint) mod() int {
	return 998244353
}

func get(i, z int) int {
	l1 := 1 * i * z
	r1 := 1 * (i + i) * z

	ans := l1

	for l1 <= r1 {
		m := (l1 + r1) / 2
		if m/z > i {
			r1 = m - 1
		} else {
			ans = max(ans, m)
			l1 = m + 1
		}
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func solve() {
	var n, m int
	fmt.Scan(&n, &m)

	dp := make([]mint, n+1)
	p := make([]mint, n+1)

	dp[n] = mint{v: 1}
	p[n] = mint{v: 1}
	for i := n - 1; i >= 1; i-- {
		dp[i].add(p[i+1])

		for z := 2; z*i <= n; z++ {
			r := get(i, z) + 1
			l := z * i
			dp[i].add(p[l])
			if r <= n {
				dp[i].subtract(p[r])
			}
		}

		p[i].add(p[i+1])
		p[i].add(dp[i])
	}

	fmt.Println(dp[1].val())
}

func main() {
	solve()
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/9f63bc7b2c354e33fad1c0c74fae046e.java
import java.util.*;

class Main {
    static final int MOD = 998244353;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        ModInt.setMod(m);

        ModInt[] dp = new ModInt[n + 1];
        ModInt[] p = new ModInt[n + 1];

        dp[n] = ModInt.valueOf(1);
        p[n] = ModInt.valueOf(1);
        for (int i = n - 1; i >= 1; i--) {
            dp[i] = dp[i].add(p[i + 1]);

            for (int z = 2; z * i <= n; z++) {
                int r = get(i, z) + 1;
                int l = z * i;
                dp[i] = dp[i].add(p[l]);
                if (r <= n) dp[i] = dp[i].subtract(p[r]);
            }

            p[i] = p[i + 1].add(dp[i]);
        }

        System.out.println(dp[1].val());
    }

    static int l1, r1;

    static int get(int i, int z) {
        l1 = 1 * i * z;
        r1 = 1 * (i + i) * z;

        int ans = l1;

        while (l1 <= r1) {
            int m = (l1 + r1) / 2;
            if (m / z > i) {
                r1 = m - 1;
            } else {
                ans = Math.max(ans, m);
                l1 = m + 1;
            }
        }

        return ans;
    }

    static class ModInt {
        static int mod;

        int value;

        ModInt(int value) {
            this.value = value;
        }

        static void setMod(int m) {
            mod = m;
        }

        static ModInt valueOf(int value) {
            return new ModInt(value % mod);
        }

        int val() {
            return value;
        }

        ModInt add(ModInt other) {
            return new ModInt((value + other.value) % mod);
        }

        ModInt subtract(ModInt other) {
            return new ModInt((value - other.value + mod) % mod);
        }

        ModInt multiply(ModInt other) {
            return new ModInt((int) ((long) value * other.value % mod));
        }

        ModInt divide(ModInt other) {
            return multiply(other.inverse());
        }

        ModInt inverse() {
            int a = value, b = mod, u = 1, v = 0;
            while (b != 0) {
                int t = a / b;
                a -= t * b;
                int temp1 = a;
                a = b;
                b = temp1;
                u -= t * v;
                int temp2 = u;
                u = v;
                v = temp2;
            }
            if (u < 0) u += mod;
            return new ModInt(u);
        }
    }
}
